triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -217,53 +220,70 @@ class Runner(
 
   private[this] def logger = ContextualizedLogger get getClass
 
-  // Handles the result of initialState or update, i.e., (s, [Commands], Text)
-  // by submitting the commands, printing the log message and returning
-  // the new state
-  private def handleStepResult(v: SValue, submit: SubmitRequest => Unit): SValue =
-    v match {
-      case SRecord(recordId, _, values)
-          if recordId.qualifiedName ==
-            QualifiedName(
-              DottedName.assertFromString("DA.Types"),
-              DottedName.assertFromString("Tuple2")) => {
-        val newState = values.get(0)
-        val commandVal = values.get(1)
-        logger.debug(s"New state: $newState")
-        commandVal match {
-          case SList(transactions) =>
-            // Each transaction is a list of commands
-            for (commands <- transactions) {
-              converter.toCommands(commands) match {
-                case Left(err) => throw new ConverterException(err)
-                case Right((commandId, commands)) => {
-                  if (usedCommandIds.contains(commandId)) {
-                    throw new RuntimeException(s"Duplicate command id: $commandId")
-                  }
-                  usedCommandIds += commandId
-                  val commandUUID = UUID.randomUUID
-                  commandIdMap += (commandUUID -> commandId)
-                  val commandsArg = Commands(
-                    ledgerId = client.ledgerId.unwrap,
-                    applicationId = applicationId.unwrap,
-                    commandId = commandUUID.toString,
-                    party = party,
-                    commands = commands
-                  )
-                  logger.debug(
-                    s"submitting command ID $commandId, commands ${commands.map(_.command.value)}")
-                  submit(SubmitRequest(commands = Some(commandsArg)))
+  import Runner.{DamlTuple2, DamlFun}
+
+  @throws[RuntimeException]
+  private def handleCommands[Z](
+      commandId: String,
+      commands: Seq[Command],
+      submit: SubmitRequest => Z): Z = {
+    if (usedCommandIds.contains(commandId)) {
+      throw new RuntimeException(s"Duplicate command id: $commandId")
+    }
+    usedCommandIds += commandId
+    val commandUUID = UUID.randomUUID
+    commandIdMap += (commandUUID -> commandId)
```

Comment:
That’s the mapping I was talking about above which we can get rid off with this change.